pkg/publisher/local/server.go

```diff
@@ -0,0 +1,87 @@
+package local
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/bacalhau-project/bacalhau/pkg/config/types"
+	"github.com/bacalhau-project/bacalhau/pkg/lib/network"
+	"github.com/rs/zerolog/log"
+)
+
+type LocalPublisherServer struct {
+	rootDirectory string
+	address       string
+	port          int
+	stopChan      chan struct{}
+}
+
+const (
+	readHeaderTimeout = 3 * time.Second
+	readTimeout       = 3 * time.Second
+)
+
+func NewLocalPublisherServer(ctx context.Context, config types.LocalPublisherConfig) *LocalPublisherServer {
+	return &LocalPublisherServer{
+		rootDirectory: config.Directory,
+		address:       resolveAddress(ctx, config.Address),
+		port:          config.Port,
+		stopChan:      make(chan struct{}),
```

Comment:
maybe we can use a sub-context with a cancel fn instead of a separate stopChan

Reply:
I tried that originally, but the context documentation actually says:

```
Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx
```

which makes it harder to keep a sub-context around.

Reply:
Actually, never mind, I got it to work as we'd expect with the context.